extensions/bitwarden/src/session.tsx

```diff
@@ -0,0 +1,307 @@
+   */
+  readonly token: string | undefined;
+
+  readonly isLoading: boolean;
+  readonly isLocked: boolean;
+  readonly isAuthenticated: boolean;
+  readonly active: boolean;
+
+  readonly repromptHash: string | undefined;
+  readonly passwordEnteredDate: Date | undefined;
+
+  /**
+   * Locks the vault.
+   */
+  readonly lock: () => Promise<void>;
+
+  /**
+   * Logs out of the vault.
+   */
+  readonly logout: () => Promise<void>;
+
+  /**
+   * Checks if the password reprompt dialog can be skipped.
+   */
+  readonly canRepromptBeSkipped: () => boolean;
+
+  /**
+   * Checks if the provided password matches the master password.
+   */
+  readonly repromptConfirmation: (password: string) => Promise<boolean>;
+}
+
+/**
+ * Component which provides a session via the {@link useSession} hook.
+ *
+ * @param props.api The Bitwarden API.
+ * @param props.unlock If true, an unlock form will be displayed if the vault is locked or unauthenticated.
+ */
+export function SessionProvider(props: PropsWithChildren<{ api: Bitwarden; unlock?: boolean }>): JSX.Element {
+  const { api, children } = props;
+  const [session, setSession] = useState<Session>({
+    api,
+    token: undefined,
+    isLoading: true,
+    isLocked: false,
+    isAuthenticated: false,
+    active: false,
+
+    repromptHash: undefined,
+    passwordEnteredDate: undefined,
+
+    lock: async function lockVault() {
+      const toast = await showToast({ title: "Locking Vault...", style: Toast.Style.Animated });
+      await api.lock();
+      await deleteToken();
+      await toast.hide();
+    },
+
+    logout: async function logoutVault() {
+      const toast = await showToast({ title: "Logging Out...", style: Toast.Style.Animated });
+      await api.logout();
+      await deleteToken();
+      await toast.hide();
+    },
+
+    canRepromptBeSkipped: function (this: Session) {
+      const { repromptIgnoreDuration } = getPreferenceValues<Preferences>();
+      if (this.passwordEnteredDate == null) {
+        return false;
+      }
+
+      const skipDuration = parseInt(repromptIgnoreDuration, 10);
+      const skipUntil = this.passwordEnteredDate.getTime() + skipDuration;
+      return Date.now() <= skipUntil;
+    },
+
+    repromptConfirmation: async function (this: Session, password: string) {
+      const providedHashed = await hashMasterPasswordForReprompting(password);
+      if (providedHashed !== this.repromptHash) {
+        return false;
+      }
+
+      const now = new Date();
+      await LocalStorage.setItem(REPROMPT_PASSWORD_ENTERED_KEY, new Date().toString());
+      setSession((old) => ({ ...old, passwordEnteredDate: now }));
+      return true;
+    },
+  });
+
+  // Functions.
+  /**
+   * Refresh the state of the session provider.
+   * This updates the component state.
+   *
+   * @param newState The state to update.
+   */
+  async function update(newState: Partial<Session>): Promise<void> {
+    const status = await api.status(newState.token);
```

Comment:
The `bw status` command is buggy in the cli, we cannot rely on it. The best approach I found out was to run the command (ex `bw list items`), and to show the form if an error occurs.

See https://github.com/bitwarden/clients/issues/2729 for more infos.

Reply:
Thanks for pointing that out! I'll change it and add a comment to capture the context.

Reply:
Context is now captured:

https://github.com/raycast/extensions/pull/3413/files#diff-59101489bd951ec97dd45c500bb43bc2775e174d64b127b034e16df3f41ec595R121-R145